Add udev semaphore synchronization for uevent-generating commands

bmr-cymru opened the pull request:
This branch adds support for udev synchronization using SysV semaphores and the device-mapper ioctl/uevent "cookie" mechanism.

The branch adds a new sub-module, `dm_udev_sync` that wraps the low-level semaphore operations and provides an interface to `core::dm`.

All udev synchronization is carried out insde `DM::do_ioctl()`. This centralises and simplifies handling ioctl errors (which require the udev semaphore to be cleaned up before propagating the error) and minimises the changes needed to synchronize with uevents.

This branch also adds basic logging infrastructure to the crate, using the log facade and the env_logger crate: this was helpful in developing and debugging the udev interactions and may be useful for debugging other problems involving device-mapper.

mulkieran commented:
@bmr-cymru Plz put up a stratisd PR that uses these changes in the manner intended. You can just edit that PRs Cargo.toml to point the devicemapper-rs dependency at this PR.

bmr-cymru commented:
Pulling this back into draft for now to fix some problems I found while testing with stratis.

bmr-cymru commented:
Rebased to current master and added a pair of commits with the changes to move `DM_UDEV_PRIMARY_SOURCE_FLAG` handling into the `UdevSync::begin()` method.

jbaublitz commented on src/core/dm_udev_sync.rs:
```diff
@@ -0,0 +1,506 @@
+    #[derive(Debug)]
+    pub struct UdevSync {
+        cookie: u32,
+        semid: i32,
```
Should this be an `Option` instead? It looks like you're using -1 as a `None` stand in.

jbaublitz commented on src/core/dm_udev_sync.rs:
```diff
@@ -0,0 +1,506 @@
+
+    use super::UdevSyncAction;
+
+    #[cfg(target_env = "musl")]
+    #[repr(C)]
+    #[derive(Debug, Default, Copy, Clone)]
+    pub struct seminfo {
+        pub semmap: ::std::os::raw::c_int,
+        pub semmni: ::std::os::raw::c_int,
+        pub semmns: ::std::os::raw::c_int,
+        pub semmnu: ::std::os::raw::c_int,
+        pub semmsl: ::std::os::raw::c_int,
+        pub semopm: ::std::os::raw::c_int,
+        pub semume: ::std::os::raw::c_int,
+        pub semusz: ::std::os::raw::c_int,
+        pub semvmx: ::std::os::raw::c_int,
+        pub semaem: ::std::os::raw::c_int,
+    }
```
I have two questions here. Why is this only for musl? I see these structures in the glibc headers too. I'm wondering if I'm misunderstanding something. Do you think it would make more sense to use bindgen here too?

bmr-cymru replied:
I'm not entirely sure why this difference exists in the `libc` crate; there is a sub-module, `gnu`, that contains `struct seminfo` (in gnu/mod.rs), and `struct semid_ds` (in gnu/b{32,64}/mod.rs). There is a separate `musl` sub-module that is compiled for musl targets that does not include these definitions:

src/unix/linux_like/linux/mod.rs
```rust
cfg_if! {
    if #[cfg(target_env = "uclibc")] {
        mod uclibc;
        pub use self::uclibc::*;
    } else if #[cfg(target_env = "musl")] {
        mod musl;
        pub use self::musl::*;
    } else if #[cfg(target_env = "gnu")] {
        mod gnu;
        pub use self::gnu::*;
    }
}
```

Looking at the commit that introduced this for `gnu`:

```
commit 775f0f20d37e56e615a88471678784ab7abf52ca
Author: David Carlier <devnexen@gmail.com>
Date:   Sun Jun 20 11:00:44 2021 +0100

    linux adding semid_ds struct and seminfo incidently.
    
    close #2002
```

And looking at the issue that it closed, rust-lang/libc#2002 it may just be an omission that the definitions are missing from musl?

This is a bit of a new type of problem for me, so I'm open to suggestions on the best way to handle it - I'm happy to file an issue to add these in the libc crate upstream. There are issues with some libc updates for musl due to the musl approach of using sanitised kernel headers (which may be out of date and lacking recent definitions - see e.g. rust-lang/libc#1571), but I don't think that will be an issue for the SysV semaphore interface.

jbaublitz reviewed (commented):
I did a first pass and have some suggestions. I want to dig a little bit deeper past basic understanding and style comments.

bmr-cymru commented on src/core/dm_udev_sync.rs:
```diff
@@ -0,0 +1,506 @@
+    }
+
+    // Missing libc SysV IPC constants from /usr/include/linux/sem.h
+    const GETVAL: i32 = 12;
+    const SETVAL: i32 = 16;
+    const SEM_INFO: i32 = 19;
```
Would that be a new crate, like devicemapper-rs-sys? I could add it to the existing bindings generation but it doesn't seem like a good fit with the ioctl / libdevmapper constants?

bmr-cymru replied:
I've added a set of commits to introduce a new `sysvsem-sys` crate, and to use bindgen for both the missing consts and (for musl builds) the two missing struct definitions.

This seems to work well and cleans up the use statements and definitions at the top of dm_udev_sync.

jbaublitz commented on src/core/dm_udev_sync.rs:
```diff
@@ -0,0 +1,459 @@
+        }
+
+        fn cancel(self) {
+            if !self.is_active() {
+                return;
+            }
+            let semid = self.semid.unwrap();
+            debug!("Canceling {:?}", self);
+            if let Err(err) = notify_sem_destroy(self.cookie, semid) {
+                error!("Failed to clean up notification semaphore: {}", err);
+            }
```
I think avoid early return when not necessary is usually better. Something like this:

```rust
if self.is_active() {
   let semid = ...
}
```

is more idiomatic.

bmr-cymru replied:
Oh I had missed this! One moment...

bmr-cymru replied:
This one should be addressed now.

jbaublitz reviewed (changes requested):
Okay, I think other than these comments, this looks great!

bmr-cymru commented on src/core/dm_udev_sync.rs:
```diff
@@ -0,0 +1,459 @@
+
+    use super::UdevSyncAction;
+
+    #[repr(C)]
+    union semun<'a> {
+        val: c_int,
+        buf: &'a semid_ds,
+        array: &'a c_ushort,
+        __buf: &'a seminfo,
+    }
```
The `semun` union is one of the weirder parts of the SysV interface - I believe originally the `semctl` function prototype used a void pointer for the 4th argument:

```
int semctl(int semid, int semnum, int cmd, void *);
```

Rather than the variadic function that is now standardised:

```
int semctl(int semid, int semnum, int cmd, ...);
```

The union is not defined in the SysV header files - rather it's shown in the manual page and standards docs that must be defined by the calling program:

```
       This function has three or four arguments, depending on cmd. 
       When there are four, the fourth has the type union semun. 
       The calling program must define this union as follows:

           union semun {
               int              val;    /* Value for SETVAL */
               struct semid_ds *buf;    /* Buffer for IPC_STAT, IPC_SET */
               unsigned short  *array;  /* Array for GETALL, SETALL */
               struct seminfo  *__buf;  /* Buffer for IPC_INFO
                                           (Linux-specific) */
           };
```

If we include that definition in the header that is passed to bindgen for the SysV pieces it produces the following Rust, which seems simpler and sufficient for our uses:

```Rust
#[repr(C)]
#[derive(Copy, Clone)]
pub union semun {
    pub val: ::std::os::raw::c_int,
    pub buf: *mut semid_ds,
    pub array: *mut ::std::os::raw::c_ushort,
    pub __buf: *mut seminfo,
}
```

bmr-cymru replied:
The reference has been eliminated now that we are using bindgen-generated versions of these types.

mulkieran commented on src/core/dm.rs:
```diff
@@ -40,6 +43,12 @@ const DM_CTL_PATH: &str = "/dev/device-mapper";
 const MIN_BUF_SIZE: usize = 16 * 1024;
 
+/// Number of device remove retry attempts
+const DM_REMOVE_RETRIES: usize = 25;
```
25 seems like so many to me; if there is a reference or any other reason for this number it's worthwile putting it here.

bmr-cymru replied:
I tend to agree - the values currently used (including the delay time) are taken directly from libdevmapper, but there doesn't appear to be any real rationale to the specific values other than to try to really make sure the removal succeeds in the face of possible interference from outside processes like udev.

In my own testing, which is the most empirical data I have for this I see the vast majority of retries succeed on the first retry, a small number on the second, and very occasionally a third or fourth so my instinct is that we can reduce this quite substantially assuming we do decide to go ahead with the retries.

This is on my list for today's standup as unfortunately the last CI failure that would require the mechanism is still unresolved.

bmr-cymru replied:
The current branch uses up to 5 retries and this seems adequate in all my testing.

mulkieran commented on src/core/dm_flags.rs:
```diff
@@ -53,23 +53,23 @@ bitflags! {
     /// for complete information about the meaning of the flags.
     #[derive(Default)]
-    pub struct DmCookie: dmi::__u16 {
+    pub struct DmUdevFlags: u32 {
```
Try dropping ```identity_op``` clippy allow, it _should_ be obsolete with this change.

bmr-cymru replied:
Gotcha, thanks! Looking into/testing this..

mulkieran reviewed (changes requested):
Two minor requests...

bmr-cymru commented on src/core/dm_udev_sync.rs:
```diff
@@ -0,0 +1,459 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
I've added additional comments to both the low level and `UdevSync` interfaces now. If they look OK to you I'll mark this as resolved?

bmr-cymru commented on src/core/dm_udev_sync.rs:
```diff
@@ -0,0 +1,459 @@
+            error!("Failed to generate udev notification semaphore: {}", err);
+            return Err(DmError::Core(errors::Error::UdevSync(err.to_string())));
+        }
+        let (cookie, semid) = cookie_pair.unwrap();
```
Done, thanks!

mulkieran commented:
@bmr-cymru Note that bindgen dependency specification is now 0.63, so your fix at https://github.com/rust-lang/rust-bindgen/pull/2247 is available. This may not change what you want to do with the source at all, but you will want to fix up the commit message.

bmr-cymru commented:
Good point. I'll update the commit to drop that mention at least - I can test removing the allowlist although in some ways I think it's desirable even with the fix for the `await` keyword. It keeps the bindings to the minimum set that we need and avoids pulling in unwanted types and interfaces.

bmr-cymru commented:
I think it's best to keep the allow listing - this avoids conflicts between the types in the generated bindings that result from the same types being exposed in both the device-mapper and SysV semaphore headers:

```

I'll edit the commit message accordingly.
error[E0428]: the name `__uid_t` is defined multiple times
    --> /home/breeves/src/git/devicemapper-rs/target/debug/build/devicemapper-sys-ab734666951c00f5/out/sem-bindings.rs:14:1
     |
14   | pub type __uid_t = ::std::os::raw::c_uint;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `__uid_t` redefined here
     |
    ::: /home/breeves/src/git/devicemapper-rs/target/debug/build/devicemapper-sys-ab734666951c00f5/out/dm-bindings.rs:1294:1
     |
1294 | pub type __uid_t = ::std::os::raw::c_uint;
     | ------------------------------------------ previous definition of the type `__uid_t` here
     |
     = note: `__uid_t` must be defined only once in the type namespace of this module

error[E0428]: the name `__gid_t` is defined multiple times
    --> /home/breeves/src/git/devicemapper-rs/target/debug/build/devicemapper-sys-ab734666951c00f5/out/sem-bindings.rs:15:1
     |
15   | pub type __gid_t = ::std::os::raw::c_uint;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `__gid_t` redefined here
     |
    ::: /home/breeves/src/git/devicemapper-rs/target/debug/build/devicemapper-sys-ab734666951c00f5/out/dm-bindings.rs:1295:1
     |
1295 | pub type __gid_t = ::std::os::raw::c_uint;
     | ------------------------------------------ previous definition of the type `__gid_t` here
     |
     = note: `__gid_t` must be defined only once in the type namespace of this module
```

mulkieran reviewed (changes requested):
Last batch of minor suggestions.

mulkieran commented:
@bmr-cymru Ok! It looks like we're good to go. Let me know when you've squashed all the commits that need to be squashed, and then I'll merge the PR and make a new release.

bmr-cymru commented:
@mulkieran I've squashed all the fixup commits into their parents and tidied the branch up a little. It's still 12 commits but I think it makes a logical sequence of changes. Let me know if you'd like me to squash more aggressively.
